File: app/src/main/java/net/xzos/UpgradeAll/server/app/engine/js/JavaScriptCoreEngine.java
package net.xzos.UpgradeAll.server.app.engine.js;

import net.xzos.UpgradeAll.json.cache.JSCacheData;
import net.xzos.UpgradeAll.server.ServerContainer;
import net.xzos.UpgradeAll.server.app.engine.api.CoreApi;
import net.xzos.UpgradeAll.server.app.engine.js.utils.JSLog;
import net.xzos.UpgradeAll.server.app.engine.js.utils.JSUtils;
import net.xzos.UpgradeAll.server.log.LogUtil;

import org.json.JSONException;
import org.json.JSONObject;
import org.mozilla.javascript.Context;
import org.mozilla.javascript.Function;
import org.mozilla.javascript.Scriptable;
import org.mozilla.javascript.ScriptableObject;

class JavaScriptCoreEngine implements CoreApi {

    private static final String TAG = "JavaScriptCoreEngine";
    private String[] LogObjectTag;
    private static final LogUtil Log = ServerContainer.AppServer.getLog();

    private String URL;
    private String jsCode;

    private Context cx;
    private Scriptable scope;

    private JSUtils JSUtils;
    private JSLog JSLog;

    JavaScriptCoreEngine(String[] logObjectTag, String URL, String jsCode) {
        this.LogObjectTag = logObjectTag;
        this.URL = URL;
        this.jsCode = jsCode;
        JSUtils = new JSUtils(this.LogObjectTag);
        JSLog = new JSLog(this.LogObjectTag);
        JSCacheData JSCacheData = new JSCacheData();
        JSUtils.setJsCacheData(JSCacheData);
    }

    String[] getLogObjectTag() {
        return LogObjectTag;
    }

    // 加载 JavaScript 代码
    private boolean executeVoidScript() {
        if (jsCode == null) return false;
        boolean isSuccess = false;
        try {
            cx.evaluateString(scope, jsCode, null, 1, null);
            isSuccess = true;
        } catch (Throwable e) {
            Log.e(LogObjectTag, TAG, String.format("executeVoidScript: 脚本载入错误, ERROR_MESSAGE: %s", e.toString()));
        }
        return isSuccess;
    }

    private boolean initRhino() {
        // 初始化 rhino 对象
        cx = Context.enter();
        cx.setOptimizationLevel(-1);
        scope = cx.initStandardObjects();
        // 载入 JavaScript 实例
        RegisterJavaMethods();
        boolean methodsSuccess = executeVoidScript();
        if (!methodsSuccess) return false;
        ScriptableObject.putProperty(scope, "URL", URL);  // 初始化 URL
        return true;
    }

    private void closeRhino() {
        Context.exit();
    }

    // 注册 Java 代码
    private void RegisterJavaMethods() {
        // 爬虫库
        Object rhinoJSUtils = Context.javaToJS(JSUtils, scope);
        ScriptableObject.putProperty(scope, "JSUtils", rhinoJSUtils);
        // Log
        Object rhinoLogUtils = Context.javaToJS(JSLog, scope);
        ScriptableObject.putProperty(scope, "Log", rhinoLogUtils);
    }

    // 运行 JS 代码
    private Object runJS(String functionName, Object[] args) {
        if (!initRhino()) return null; // 初始化 J2V8
        Function function = (Function) scope.get(functionName, scope);
        Object result = function.call(cx, scope, scope, args);
        closeRhino(); // 销毁 J2V8 对象
        return result;
    }

    @Override
    public String getDefaultName() {
        String defaultName = Context.toString(runJS("getDefaultName", new Object[]{}));
        Log.d(LogObjectTag, TAG, "getDefaultName: defaultName: " + defaultName);
        return defaultName;
    }

    @Override
    public int getReleaseNum() {
        int releaseNum = (int) Context.toNumber(runJS("getReleaseNum", new Object[]{}));
        Log.d(LogObjectTag, TAG, "getReleaseNum: releaseNum: " + releaseNum);
        return releaseNum;
    }


    @Override
    public String getVersioning(int releaseNum) {
        Object result;
        Object[] args = {releaseNum};
        try {
            result = runJS("getVersioning", args);
        } catch (ClassCastException e) {
            // TODO: 向下兼容两个主版本后移除，当前版本：0.1.0-alpha.3
            Log.w(LogObjectTag, TAG, "getVersioning: 未找到 getVersioning 函数，尝试向下兼容");
            result = runJS("getVersionNumber", args);
        }
        String versionNumber = Context.toString(result);
        Log.d(LogObjectTag, TAG, "getVersioning: versionNumber: " + versionNumber);
        return versionNumber;
    }

    @Override
    public String getChangelog(int releaseNum) {
        Object[] args = {releaseNum};
        String changeLog = Context.toString(runJS("getChangelog", args));
        Log.d(LogObjectTag, TAG, "getChangelog: Changelog: " + changeLog);
        return changeLog;
    }

    @Override
    public JSONObject getReleaseDownload(int releaseNum) {
        Object[] args = {releaseNum};
        String fileJsonString = Context.toString(runJS("getReleaseDownload", args));
        JSONObject fileJson = new JSONObject();
        try {
            fileJson = new JSONObject(fileJsonString);
        } catch (JSONException e) {
            Log.e(LogObjectTag, TAG, "getReleaseDownload: 返回值不符合 JsonObject 规范, fileJsonString : " + fileJsonString);
        } catch (NullPointerException e) {
            Log.e(LogObjectTag, TAG, "getReleaseDownload: 返回值为 NULL, fileJsonString : " + fileJsonString);
        }
        Log.d(LogObjectTag, TAG, "getReleaseDownload:  fileJson: " + fileJson);
        return fileJson;
    }
}

File: app/src/main/kotlin/net/xzos/upgradeAll/server/app/engine/js/JavaScriptCoreEngine.kt
package net.xzos.upgradeAll.server.app.engine.js

import com.google.gson.Gson
import com.google.gson.JsonSyntaxException
import kotlinx.coroutines.runBlocking
import net.xzos.upgradeAll.data.json.gson.JSReturnData
import net.xzos.upgradeAll.server.ServerContainer
import net.xzos.upgradeAll.server.app.engine.js.utils.JSLog
import net.xzos.upgradeAll.server.app.engine.js.utils.JSUtils
import org.mozilla.javascript.Context
import org.mozilla.javascript.Function
import org.mozilla.javascript.ScriptableObject

internal class JavaScriptCoreEngine(
        private val logObjectTag: Pair<String, String>,
        private val URL: String?,
        private val jsCode: String?
) : JavaScriptApi {

    private val jsLog = JSLog(logObjectTag)
    internal lateinit var jsUtils: JSUtils
    private lateinit var scope: ScriptableObject

    init {
        runBlocking { runJS(null, arrayOf()) }  // 空运行初始化基础 JavaScript 运行环境
    }

    private fun initRhino(cx: Context, scope: ScriptableObject) {
        // 初始化 rhino 对象
        cx.optimizationLevel = -1
        // 载入 JavaScript 实例
        registerJavaMethods(cx, scope)
        val methodsSuccess = executeVoidScript(cx, scope)
        if (methodsSuccess)
            ScriptableObject.putProperty(scope, "URL", URL)  // 初始化 URL
    }

    // 注册 Java 代码
    private fun registerJavaMethods(cx: Context, scope: ScriptableObject) {
        // 爬虫库
        if (!::jsUtils.isInitialized)
            jsUtils = JSUtils(logObjectTag, scope)
        val rhinoJSUtils = Context.javaToJS(jsUtils.get(cx), scope)
        ScriptableObject.putProperty(scope, "JSUtils", rhinoJSUtils)
        // Log
        val rhinoLogUtils = Context.javaToJS(jsLog, scope)
        ScriptableObject.putProperty(scope, "Log", rhinoLogUtils)
    }

    // 加载 JavaScript 代码
    private fun executeVoidScript(cx: Context, scope: ScriptableObject): Boolean {
        if (jsCode == null) return false
        return try {
            cx.evaluateString(scope, jsCode, logObjectTag.toString(), 1, null)
            true
        } catch (e: Throwable) {
            Log.e(logObjectTag, TAG, String.format("executeVoidScript: 脚本载入错误, ERROR_MESSAGE: %s", e.toString()))
            false
        }
    }

    // 运行 JS 代码
    private fun runJS(functionName: String?, args: Array<Any>): Any? {
        val cx = Context.enter()
        val scope = cx.initStandardObjects().also {
            initRhino(cx, it)
            scope = it
        }
        return try {
            if (functionName != null) {
                (scope.get(functionName, scope) as Function).call(cx, scope, scope, args)
            } else null
        } catch (e: Throwable) {
            Log.e(logObjectTag, TAG, "runJS: 脚本执行错误, 函数名: $functionName, ERROR_MESSAGE: $e")
            null
        } finally {
            Context.exit()
        }
    }

    /**
     * 返回更新项默认名称
     */
    override suspend fun getDefaultName(): String? {
        val result = runJS("getDefaultName", arrayOf()) ?: return null
        val defaultName = Context.toString(result)
        Log.d(logObjectTag, TAG, "getDefaultName: defaultName: $defaultName")
        return defaultName
    }

    /**
     * 返回更新项默认图标 URL
     */
    override suspend fun getAppIconUrl(): String? {
        val result = runJS("getAppIconUrl", arrayOf())
                ?: return null
        val appIconUrl = Context.toString(result)
        Log.d(logObjectTag, TAG, "getAppIconUrl: appIconUrl: $appIconUrl")
        return appIconUrl
    }

    /**
     * 返回由 JavaScript 函数返回的固定 JSON 格式生成的版本信息数据类
     */
    override suspend fun getReleaseInfo(): JSReturnData {
        val result = runJS("getReleaseInfo", arrayOf())
                ?: return JSReturnData()
        val versionInfoJsonString: String = Context.toString(result)
        Log.d(logObjectTag, TAG, "getReleaseInfo: JSON 字符串: $versionInfoJsonString")
        return try {
            JSReturnData(
                    Gson().fromJson(versionInfoJsonString, Array<JSReturnData.ReleaseInfoBean>::class.java).also {
                        Log.d(logObjectTag, TAG, "getReleaseInfo: JSON 解析成功")
                    }.toList()
            )
        } catch (e: JsonSyntaxException) {
            Log.e(logObjectTag, TAG, "getReleaseInfo: JSON 解析失败, ERROR_MESSAGE: $e")
            JSReturnData()
        }
    }

    /**
     * 下载文件操作
     * 操作成功返回 true
     */
    override suspend fun downloadReleaseFile(downloadIndex: Pair<Int, Int>): Boolean =
            runJS("downloadReleaseFile", arrayOf(downloadIndex.first, downloadIndex.second)) != null

    companion object {
        private const val TAG = "JavaScriptCoreEngine"
        private val Log = ServerContainer.Log
    }

}

private interface JavaScriptApi {
    suspend fun getDefaultName(): String?
    suspend fun getAppIconUrl(): String?
    suspend fun getReleaseInfo(): JSReturnData?
    suspend fun downloadReleaseFile(downloadIndex: Pair<Int, Int>): Boolean
}
